src/version.rs

```diff
@@ -0,0 +1,934 @@
+#[derive(Clone, Debug, Default, Eq, PartialEq)]
+pub struct NEVRA<'a> {
+    name: Cow<'a, str>,
+    evr: EVR<'a>,
+    arch: Cow<'a, str>,
+}
+
+impl<'a> NEVRA<'a> {
+    /// Create a new NEVRA
+    pub fn new<T: Into<Cow<'a, str>>>(
+        name: T,
+        epoch: T,
+        version: T,
+        release: T,
+        arch: T,
+    ) -> NEVRA<'a> {
+        Self {
+            name: name.into(),
+            evr: EVR::new(epoch, version, release),
+            arch: arch.into(),
+        }
+    }
+
+    /// Create a NEVRA parsed from a string
+    pub fn parse(nevra: &'a str) -> Self {
+        let (n, e, v, r, a) = NEVRA::parse_values(nevra);
+        Self::new(n, e, v, r, a)
+    }
+
+    /// The name value
+    pub fn name(&self) -> &str {
+        &self.name
+    }
+
+    /// The EVR
+    pub fn evr(&'a self) -> &'a EVR<'a> {
+        &self.evr
+    }
+
+    /// The epoch value
+    pub fn epoch(&self) -> &str {
+        &self.evr.epoch
+    }
+
+    /// The version value
+    pub fn version(&self) -> &str {
+        &self.evr.version
+    }
+
+    /// The release value
+    pub fn release(&self) -> &str {
+        &self.evr.release
+    }
+
+    /// The arch value
+    pub fn arch(&self) -> &str {
+        &self.arch
+    }
+
+    /// Return the epoch, version and release values as a 5-element tuple
+    pub fn values(&self) -> (&str, &str, &str, &str, &str) {
+        (
+            &self.name,
+            &self.evr.epoch,
+            &self.evr.version,
+            &self.evr.release,
+            &self.arch,
+        )
+    }
+
+    /// Parse the name, epoch, version, release and arch values and return them as a 5-element tuple
+    pub fn parse_values(nevra: &'a str) -> (&'a str, &'a str, &'a str, &'a str, &'a str) {
+        let (name, evra) = nevra.split_once('-').unwrap_or((nevra, ""));
+        let (epoch, vra) = evra.split_once(':').unwrap_or(("", evra));
+        let (version, ra) = vra.split_once('-').unwrap_or((vra, ""));
+        let (release, arch) = ra.rsplit_once('.').unwrap_or((ra, ""));
+
+        (name, epoch, version, release, arch)
+    }
+
+    /// Write an NEVRA string in a normalized form which always includes an epoch
+    ///
+    /// The standard string representation of an EVR will ignore the epoch if not set.  A package
+    /// having no epoch value is equivalent to having an epoch of zero. Sometimes it is useful to
+    /// write NEVRAs in a form such that equivalent values are represented identically, therefore,
+    /// this "normalized" form will always include it.
+    pub fn as_normalized_form(&self) -> String {
+        format!(
+            "{}-{}.{}",
+            self.name,
+            self.evr.as_normalized_form(),
+            self.arch
+        )
+    }
+}
+
+impl<'a> fmt::Display for NEVRA<'a> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{}-{}.{}", self.name, self.evr, self.arch)
```

Comment:
Is there a particular reason we are not using the normalized form, always?

Reply:
The short form is actually more common, it's only in specific places like repo metadata where the normalized form ends up being used.

Reply:
The current implementation works the same way that `rpm -qp --queryformat="%{NEVRA}" $pkg` does